syntax = "proto3";

package com.targist.ardui.proto;

message GenericArduinoProgram {
  SetupScript setup = 1;
  LoopScript loop = 2;
}

message SetupScript { repeated Instruction instructions = 1; }

message LoopScript { repeated Instruction instructions = 1; }

message Instruction {
  oneof action {
    SetPinMode setPinMode = 1;
    DigitalWrite digitalWrite = 2;
    AnalogWrite analogWrite = 3;
    Sleep sleep = 4;
  }
}

enum Mode {
  INPUT = 0;
  OUTPUT = 1;
  INPUT_PULLUP = 2;
}

enum Level {
  HIGH = 0;
  LOW = 1;
}

message SetPinMode {
  Mode mode = 1;
  uint32 pin = 2;
}

message DigitalWrite {
  Level level = 1;
  uint32 pin = 2;
}

message AnalogWrite {
  uint32 pin = 1;
  // Duty cycle which should be between 0 and 255
  uint32 value = 2;
}

message Sleep { uint32 duration = 1; }
